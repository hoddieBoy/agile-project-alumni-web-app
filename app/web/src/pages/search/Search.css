body {
    background-color: #f8f9fa;
    margin: 0;
}

main {
    flex: 1;
}

.text-custom-primary {
    color: #002855 !important;
}

.text-custom-secondary {
    color: #6c757d !important;
}

.card {
    border: 1px solid #002855;
}

.card-body {
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.card-title {
    margin: 0;
}

.card-text {
    margin: 0;
}

.btn-primary {
    background-color: #002855 !important;
    border-color: #002855 !important;
}

.btn-outline-primary {
    border-color: #002855 !important;
    color: #002855 !important;
}

.btn-outline-primary:hover {
    background-color: #002855 !important;
    color: #ffffff !important;
}

label,
.form-control {
    color: #002855 !important;
}

.btn-custom-primary {
    background-color: #002855 !important;
    border-color: #002855 !important;
}
