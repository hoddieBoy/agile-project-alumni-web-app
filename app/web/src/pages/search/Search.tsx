import React, {useEffect, useState} from 'react';
import {AlumniData} from 'payload/response/AlumniData';
import Header from 'components/Header';
import Footer from 'components/Footer';
import {Spin} from 'antd';
import {Form, useActionData} from "react-router-dom";

type SearchCriteria = {
    name: string;
    graduationYear: string;
    currentCompany: string;
    city: string;
};

type ErrorMessage = {
    message: string;
};

function Search(): React.ReactElement {
    const [isLoading, setIsLoading] = useState(false);
    const [searchCriteria, setSearchCriteria] = useState<SearchCriteria>({
        name: '',
        graduationYear: '',
        currentCompany: '',
        city: ''
    });

    const data = useActionData() as AlumniData[] | ErrorMessage;

    useEffect(() => {
        setIsLoading(false);
    }, [data]);

    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
        const {id, value} = e.target;
        setSearchCriteria((prev) => ({...prev, [id]: value}));
    };

    const handleReset = () => {
        setSearchCriteria({
            name: '',
            graduationYear: '',
            currentCompany: '',
            city: ''
        });
    }

    const handleSearch = async () => {
        setIsLoading(true);
    }

    return (
        <>
            <Header/>
            <main className="container my-4">
                <h2 className="text-center text-custom-primary">Find Alumni</h2>
                <p className="text-center text-custom-secondary">
                    Search for specific alumni based on various criteria
                </p>

                <Form
                    className="text-center border p-4 rounded my-4"
                    action="/search"
                    method="post"
                    onSubmit={handleSearch}
                >
                    <div className="row mb-4">
                        <div className="form-group col-md-3">
                            <label htmlFor="name" className="form-label">Nom</label>
                            <input
                                type="text"
                                className="form-control"
                                id="name"
                                name="name"
                                placeholder="Entrez un nom"
                                value={searchCriteria.name}
                                onChange={handleInputChange}
                            />
                        </div>
                        <div className="form-group col-md-3">
                            <label htmlFor="graduationYear" className="form-label">Année De Diplomation</label>
                            <input
                                type="text"
                                className="form-control"
                                id="graduationYear"
                                name="graduationYear"
                                placeholder="Entrez une année"
                                value={searchCriteria.graduationYear}
                                onChange={handleInputChange}
                            />
                        </div>
                        <div className="form-group col-md-3">
                            <label htmlFor="currentCompany" className="form-label">Entreprise Actuelle</label>
                            <input
                                type="text"
                                className="form-control"
                                id="currentCompany"
                                name="currentCompany"
                                placeholder="Entrez une entreprise"
                                value={searchCriteria.currentCompany}
                                onChange={handleInputChange}
                            />
                        </div>
                        <div className="form-group col-md-3">
                            <label htmlFor="city" className="form-label">Ville</label>
                            <input
                                type="text"
                                className="form-control"
                                id="city"
                                name="city"
                                placeholder="Entrez une ville"
                                value={searchCriteria.city}
                                onChange={handleInputChange}
                            />
                        </div>
                    </div>
                    <div className="row justify-content-center gap-3">
                        <button
                            type="reset"
                            className="btn btn-outline-secondary form-group col-md-2"
                            onClick={handleReset}
                        >
                            Reinitialiser
                        </button>
                        <button
                            type="submit"
                            className="custom-primary-color btn btn-primary btn-custom-primary form-group col-md-2"
                        >
                            Rechercher
                        </button>
                    </div>
                </Form>

                <section className="scrollable-container">
                    {isLoading ? (
                        <div className="d-flex justify-content-center my-4">
                            <Spin size="large"/>
                        </div>
                    ) : (!data || 'message' in data) ? (
                        <p className="text-center text-danger">{data?.message}</p>
                    ) : (
                        data.map((alumni) => (
                            <div className="card mb-3" key={alumni.id}>
                                <div className="card-body d-flex align-items-center justify-content-between">
                                    <div className="d-flex align-items-center">
                                        <img
                                            src="https://static.wikia.nocookie.net/be-like-bro/images/6/6d/Anon-Man-1-.jpg/revision/latest?cb=20170605170350"
                                            alt="Profile"
                                            className="rounded-circle me-3"
                                            style={{height: '100px', width: '100px', objectFit: 'cover'}}
                                        />
                                        <div>
                                            <h5 className="card-title">{alumni.full_name}</h5>
                                            <p className="card-text">{`Graduated in ${alumni.graduation_year}, ${alumni.current_company}, ${alumni.city}`}</p>
                                        </div>
                                    </div>
                                    <a href="#" className="btn btn-outline-primary">View Profile</a>
                                </div>
                            </div>
                        ))
                    )}
                </section>
            </main>
            <Footer/>
        </>
    );
}

export default Search;