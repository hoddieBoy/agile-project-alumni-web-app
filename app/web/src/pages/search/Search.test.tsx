import React from "react";
import Search from "pages/search/Search";
import {createMemoryRouter, RouterProvider} from "react-router-dom";
import {render, screen, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SearchAction from "pages/search/Search.action";

describe("The Search page functionality", () => {
    const router = createMemoryRouter(
        [
            {
                path: '/search',
                element: <Search/>,
                action: SearchAction
            },
        ],
        {
            initialEntries: ['/search']
        }
    );

    describe("Render and basic elements", () => {
        test("The Search page renders without crashing", () => {
            const {container} = render(<RouterProvider router={router}/>);
            expect(container).toMatchSnapshot();
        });

        test("The Search page contains a form with the correct fields", () => {
            render(<RouterProvider router={router}/>);
            expect(screen.getByLabelText(/name/i)).toBeInTheDocument();
            expect(screen.getByLabelText(/city/i)).toBeInTheDocument();
        });

        test("The Search page contains a submit button", () => {
            render(<RouterProvider router={router}/>);
            expect(screen.getByRole('button', {name: /search/i})).toBeInTheDocument();
        });
    });

    describe("Given a user on the Search page", () => {
        afterEach(() => {
            jest.restoreAllMocks();
        });

        test("When the user submits the form without entering any data, then an error message is displayed", async () => {
            render(<RouterProvider router={router}/>);
            userEvent.click(screen.getByRole('button', {name: /search/i}));
            await waitFor(() => expect(screen.getByText('At least one search criteria is required.')).toBeInTheDocument());
        });
    });
});