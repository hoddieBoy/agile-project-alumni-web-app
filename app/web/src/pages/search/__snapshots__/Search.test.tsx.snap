// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Search page functionality Render and basic elements The Search page renders without crashing 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="header-left"
    >
      <a
        aria-label="IMT Atlantique"
        class="navbar-brand"
        href="/"
      >
        <img
          alt="IMT Atlantique Logo"
          src="https://www.imt-atlantique.fr/sites/default/files/Images/Ecole/charte-graphique/IMT_Atlantique_logo_RVB_Negatif_Baseline_400x272.png"
        />
      </a>
      <input
        aria-label="Search"
        class="form-control navbar-search"
        placeholder="Search"
        type="text"
      />
    </div>
    <nav
      class="header-right gap-3"
    >
      <a
        class="nav-item nav-link"
        href="/search"
      >
        Search
      </a>
      <a
        class="nav-item nav-link"
        href="/import-export"
      >
        Import-Export
      </a>
      <a
        class="nav-item nav-link"
        href="/gestion-utilisateurs"
      >
        Gestion Utilisateurs
      </a>
      <a
        class="nav-item nav-link"
        href="/stat"
      >
        Stat
      </a>
      <button
        class="btn btn-primary"
        type="button"
      >
        Log out
      </button>
    </nav>
  </header>
  <main
    class="container my-4"
  >
    <h2
      class="text-center text-custom-primary"
    >
      Find Alumni
    </h2>
    <p
      class="text-center text-custom-secondary"
    >
      Search for specific alumni based on various criteria
    </p>
    <form
      action="/search"
      class="text-center border p-4 rounded my-4"
      method="post"
    >
      <div
        class="row mb-4"
      >
        <div
          class="form-group col-md-3"
        >
          <label
            class="form-label"
            for="name"
          >
            Name
          </label>
          <input
            class="form-control"
            id="name"
            name="name"
            placeholder="Enter name"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group col-md-3"
        >
          <label
            class="form-label"
            for="graduationYear"
          >
            Graduation Year
          </label>
          <select
            class="form-control"
            id="graduationYear"
            name="graduationYear"
          >
            <option
              value=""
            >
              Select Year
            </option>
            <option
              value="2024"
            >
              2024
            </option>
            <option
              value="2023"
            >
              2023
            </option>
            <option
              value="2022"
            >
              2022
            </option>
            <option
              value="2021"
            >
              2021
            </option>
            <option
              value="2020"
            >
              2020
            </option>
            <option
              value="2019"
            >
              2019
            </option>
            <option
              value="2018"
            >
              2018
            </option>
            <option
              value="2017"
            >
              2017
            </option>
            <option
              value="2016"
            >
              2016
            </option>
            <option
              value="2015"
            >
              2015
            </option>
          </select>
        </div>
        <div
          class="form-group col-md-3"
        >
          <label
            class="form-label"
            for="currentCompany"
          >
            Current Company
          </label>
          <input
            class="form-control"
            id="currentCompany"
            name="currentCompany"
            placeholder="Enter company name"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group col-md-3"
        >
          <label
            class="form-label"
            for="city"
          >
            City
          </label>
          <input
            class="form-control"
            id="city"
            name="city"
            placeholder="Enter city"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="row justify-content-center gap-3"
      >
        <button
          class="btn btn-outline-secondary form-group col-md-2"
          type="reset"
        >
          Reset
        </button>
        <button
          class="custom-primary-color btn btn-primary btn-custom-primary form-group col-md-2"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
    <section
      class="scrollable-container"
    >
      <p
        class="text-center text-danger"
      />
    </section>
  </main>
  <footer>
    <div
      class="container"
    >
      <span>
        Â© 2024 Alumni FIL
      </span>
    </div>
  </footer>
</div>
`;
